managers/asset-operations-manager.js

```diff
@@ -398,60 +431,331 @@ class AssetOperationsManager {
+                getPublicOperationResult.data = {
+                    errorType: 'DKG_CLIENT_ERROR',
+                    errorMessage: "Calculated root hashes don't match!",
+                };
+
+                // TODO: Check returned response
+                return {
+                    UAL,
+                    getPublicOperationResult,
+                };
+            }
+
+            // Transform public assertion to include updated private assertion Id
+            const hashRegex = /"([^"]+)"/; // Regex pattern to match the hash value inside the double quotes
+            publicAssertion = publicAssertion.map((str) => {
+                if (str.includes('privateAssertionID')) {
+                    return str.replace(hashRegex, `"${privateAssertionId}"`);
+                }
+                return str;
+            });
+        } else {
+            const publicGraph = {
+                '@graph': [
+                    jsonContent.public && !isEmptyObject(jsonContent.public)
+                        ? jsonContent.public
+                        : null,
+                    jsonContent.private && !isEmptyObject(jsonContent.private)
+                        ? {
+                              [PRIVATE_ASSERTION_PREDICATE]: privateAssertionId,
+                          }
+                        : null,
+                ],
+            };
+            publicAssertion = await formatAssertion(publicGraph);
+        }
+
+        const publicAssertionId = calculateRoot(publicAssertion);
+
+        const contentAssetStorageAddress = await this.blockchainService.getContractAddress(
+            'ContentAssetStorage',
+            blockchain,
+        );
+
+        let tokenAmountInWei;
+
+        if (tokenAmount != null) {
+            tokenAmountInWei = tokenAmount;
+        } else {
+            const assetStorageContractAddress = resolveUAL(UAL).contract;
+            const firstAssertionId = await this.blockchainService.getAssertionIdByIndex(
+                tokenId,
+                0,
+                blockchain,
+            );
+
+            const keyword = ethers.solidityPacked(
+                ['address', 'bytes32'],
+                [assetStorageContractAddress, firstAssertionId],
+            );
+
+            const agreementId = ethers.sha256(
+                ethers.solidityPacked(
+                    ['address', 'uint256', 'bytes'],
+                    [assetStorageContractAddress, tokenId, keyword],
+                ),
+            );
+            const agreementData = await this.blockchainService.getAgreementData(
+                agreementId,
+                blockchain,
+            );
+
+            const now = await this.blockchainService.getBlockchainTimestamp(blockchain);
+            const currentEpoch = Math.floor(
+                (now - agreementData.startTime) / agreementData.epochLength,
+            );
+
+            const epochsLeft = agreementData.epochsNumber - currentEpoch;
```

Comment:
State can be finalized in the current epoch and we should always have a buffer here (include current epoch) imo. Let's check if we have the same mechanism on the node side (for bid validation)